tests/test_kmk_keys.py

```diff
@@ -0,0 +1,87 @@
+import unittest
+
+from kmk.keys import KC, Key, ModifierKey
+from tests.keyboard_test import KeyboardTest
+
+
+class TestKmkKeys(unittest.TestCase):
+    def test_basic_kmk_keyboard(self):
+        keyboard = KeyboardTest(
+            [],
+            [
+                [
+                    KC.HASH,
+                    KC.RALT(KC.HASH),
+                    KC.RALT(KC.LSFT)(KC.N3),
```

Comment:
Not a fan of that syntax, but at least `KC.RALT(KC.LSFT(KC.N3))` passes the same test.
No need to necessarily change it.

Reply:
This weird syntax here is intentional, I was trying to exercise a particular edge case. These three lines _should_ be synonymous:

- KC.RALT(KC.HASH)
- KC.RALT(KC.LSFT(KC.N3))
- KC.RALT(KC.LSFT)(KC.N3)

But, the third one actually fails on the current master branch. Maybe I should have added a comment this effect inline.

Reply:
I fail to see how the third line should not fail. The syntax looks invalid, and even if it's not, it's confusing, and should be avoided IMHO

Reply:
This this valid?
````
>>> KC.RALT(KC.LSFT)
ModifierKey(code=-1, has_modifiers={64, 2})
````
If so, what should this do?
````
>>> ALT_SHIFT = KC.RALT(KC.LSFT)
>>> ALT_SHIFT(KC.N3)
````
I'd argue that it should produce:
`Key(code=32, has_modifiers={64, 2})`
Currently it produces:
`Key(code=32, has_modifiers={-1})`
Which I think is pretty heavily broken.

I'm not arguing it's a good idea, I was just writing a test that exercises an edge case that was producing an undesirable result. Nevertheless, I'm happy to change the test if you'd like.

Reply:
```
>>> ALT_SHIFT = KC.RALT(KC.LSFT)
>>> ALT_SHIFT(KC.N3)
```
This should fail I believe. The reason that `KC.RALT(KC.HASH)` this will pass is because we handled the cases for shifted keys by design. `KC.RALT(KC.LSFT(KC.N3))` This will work because mod+key (and therefor shifted keys) and keys passed to other keys. Meanwhile `KC.RALT(KC.LSFT)(KC.N3)` this is a key, with another key passed to it, and then just an arbitrary extra (), which seems syntactically incorrect at best. I'm not saying that we can't add support for something like your second example in some way, but the test case I noted should fail, seems like it would fail correctly, and we should handle what your intent is differently under the hood. If you feel I'm wrong, please explain why. I rarely claim to be correct, but I am questioning why you think that syntax is valid, not the idea of what you showed in `ALT_SHIFT(KC.N3)` this example. The idea, to be clear, is valid.

Reply:
Which parentheses are arbitrary?
`A(B)(C)`
is equivalent to
```
X = A(B)
X(C)
```
`A(B)C` on the other hand wouldn't even parse.

In the case here…
`KC.RALT(KC.LSFT)(KC.N3)` is equivalent to:
```
X = KC.RALT(KC.LSFT)
X(KC.N3)
```

Whether it is on one line or two, this code does the following:
1. `KC.RALT(KC.LSFT)` returns a ModifierKey with both mods in has_modifiers. I assumed that this is correct/intentional. If this behavior is wrong or unintended, then I can understand the confusion. This case seems to be handled intentionally in the code, as the special case of setting the keycode to -1 is specific to this situation (calling a ModifierKey w/ another ModifierKey as the parameter)
2. The ModifierKey returned from `KC.RALT(KC.LSFT)` is called with KC.N3 as the parameter. If this does anything at all, it should return the parameter key with the content of the ModifierKey's has_modifiers added to it. Instead, it only adds the `-1` fake keycode.

I mean, do we agree that returning `Key(code=32, has_modifiers={-1})` here is undesirable?

Reply:
TIL. Yeah, you are correct. Thanks for clearing that up.

Reply:
Of course, and thanks for reviewing the PR. This part of the code was/is very confusing. There is clearly a reason for the `# FIXME this is atrocious to read…` note in the section I replaced.